### PR TITLE
[libcxx] Use generic builtins for popcount, clz and ctz

### LINKED ISSUE #86556: [libcxx] utilize __builtin_popcountg, __builtin_clzg, __builtin_ctzg
clang-19+ and gcc-14+ will support type generic builtins (only for unsigned types) for these builtins.

I see them being used in:
- libcxx/include/__bit/countl.h (__builtin_clz{l|ll})
- libcxx/include/__bit/countr.h (__builtin_ctz{l|ll})
- src/include/ryu/ryu.h (ctz)
- include/__bit/popcount.h (__builtin_popcount{l|ll})
- test/libcxx/containers/unord/next_pow2.pass.cpp (popcount)

There's currently a slight issue in clang-19 with the clz + ctz builtins that I expect to get resolved quickly and before the clang-19 release.  https://github.com/llvm/llvm-project/issues/86549

(Consider whether this request may improve things, and if so, does it make sense to tag it as a good first issue?)

https://github.com/llvm/llvm-project/commit/b1e40ede9fd2cd33c7f533147827b23541c516c9 is an example of me utilizing `__builtin_popcountg` in llvm-libc as an example.

### PR DESCRIPTION
**Pull request by marcauberer**
Fixes #86556

Use `__builtin_popcountg` instead of `__buildin_popcount{l|ll}`
Use `__builtin_clzg instead` of `__buildin_clz{l|ll}`
Use `__builtin_ctzg instead` of `__builtin_ctz{l|ll}`

The generic variant of the builtins can be used to simplify some logic with >= Clang 19 or >= GCC 14, where these generic variants are available. As for backwards compatibility reasons, we can't completely remove the old logic. Therefore, I left ToDo comments to address this, as soon as support for pre Clang 19 as well as pre GCC 14 is dropped.

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
Right, so these builtins are nascent. Only unreleased versions of clang and gcc support them (top of tree).  So you'll need to use `__has_builtin` to detect support for them. `__has_builtin` also might not exist in older compiler versions, so you may need to check that.  Try to find what compiler versions libcxx currently supports; your solution will need to consider those older compilers which may lack the builtins.

File: libcxx/include/__bit/countl.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countl.h:64]: `__builtin_clzg` can handle `__uint128_t` I suspect.  No need to do any of this.

**Review by nickdesaulniers — COMMENTED** [review]
Consider that if the compiler has support for that builtin, `__countl_zero` itself could be implemented almost entirely in terms of a call to `__builtin_clzg`.

Same for `__countr_zero` / `__builtin_ctzg`, and `popcount`/ `__builtin_popcountg`.

**Comment by philnik777** [comment]
What benefit does this change have?

**Comment by nickdesaulniers** [comment]
> What benefit does this change have?

I suspect it would be possible to omit most of the content of libcxx/include/__bit/countl.h, libcxx/include/__bit/countr.h, and libcxx/include/__bit/popcount.h when this builtin is defined.

The current PR is perhaps too granular in its use of __has_builtin checks.  For example, take the `__libcpp_clz` case. `__libcpp_clz` need not even be defined when the type generic builtin exists. Callers of __libcpp_clz could be updated to just call `__countl_zero`.

That would allow the preprocessor/lexer and Sema to do less work when these builtins exist.  Is that worth the change here? I'll leave that to libc++ maintainers to make a call.

File: libcxx/include/__bit/countl.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countl.h:79]: I think it would be simpler (and more obvious to fellow reviewers) to do something like:
```diff
diff --git a/libcxx/include/__bit/countl.h b/libcxx/include/__bit/countl.h
index 396cfc2c3f40..5951bc1cc9f9 100644
--- a/libcxx/include/__bit/countl.h
+++ b/libcxx/include/__bit/countl.h
@@ -24,6 +24,14 @@ _LIBCPP_PUSH_MACROS
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
+#if __has_builtin(__builtin_clzg)
+template <class _Tp>
+_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 int __countl_zero(_Tp __t) _NOEXCEPT {
+static_assert(__libcpp_is_unsigned_integer<_Tp>::value, "__countl_zero requires an unsigned integer type");
+  return __builtin_clzg(__x, numeric_limits<_Tp>::digits);
+}
+#else
+
 _LIBCPP_NODISCARD inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR int __libcpp_clz(unsigned __x) _NOEXCEPT {
   return __builtin_clz(__x);
 }
@@ -80,6 +88,7 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 int __countl_zero(_Tp __t) _
     return __ret + __iter;
   }
 }
+#endif // __has_builtin(__builtin_clzg)
 
 #if _LIBCPP_STD_VER >= 20
```
then update callsites of `__libcpp_clz` to just use `__countl_zero` instead.

If updating the callsites is untenable, then the range of that `#if __has_builtin(__builtin_clzg) can be tightened up to just skip the existing definition of __countl_zero (and not touch definitions of `__libcpp_clz`).

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Done.

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Implemented the second option. Would have been too many callsites to change, but maybe this can be done in a follow-up.

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Done.

**Comment by marcauberer** [comment]
The build failure should be fixed with #86577, so let's wait for it to land.

File: libcxx/include/__bit/countl.h
Comment (mordante) [inline libcxx/include/__bit/countl.h:32]: I don't see a real win here. If we can use `__builtin_clzg` unconditionally in a few years we might be able to do more refactoring to remove several of these overloads.

File: libcxx/include/__bit/countl.h
Comment (mordante) [inline libcxx/include/__bit/countl.h:64]: I feel this is a win.
Please move this code above the comment. The comment is clearly intended for the `#else` branch.

File: libcxx/include/__bit/countl.h
Comment (mordante) [inline libcxx/include/__bit/countl.h:79]: This looks wrong
```suggestion
  return __builtin_clzg(__t) - (numeric_limits<unsigned>::digits - numeric_limits<_Tp>::digits);
                                                                              ~~~~~~~~
```

This is the kind of cleanups I would indeed like with this patch.

File: libcxx/src/include/ryu/d2s_intrinsics.h
Comment (mordante) [inline libcxx/src/include/ryu/d2s_intrinsics.h:252]: Please undo the changes to the ryu files. This is code used from an external project. I like to keep the deviation from the original source minimal.

File: libcxx/include/__bit/countl.h
Comment (mordante) [inline libcxx/include/__bit/countl.h:28]: Please add comment at the top of the file with a TODO explaining that `__builtin_clzg` is available with
starting clang-19 and gcc-14. That should help is to know when we can enable the new builtins unconditionally.

**Review by mordante — COMMENTED** [review]
Thanks for your patch. I like to only use the new builtin where it carries its weight.

Please explain in the commit message why this change is useful.

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Done. To all three files.

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Yes, I agree. I also was not very convinced by using the generic variant by any means. Removed the `#ifs` in the abstraction helpers.

**Review by mordante — CHANGES_REQUESTED** [review]
I've one minor comment, other than that it LGTM.

There is one issue; our CI is not testing with a Clang version that has https://github.com/llvm/llvm-project/pull/86577. I will look at that and let you know when the CI is ready to really test this patch. Hence the "Request changes" status.

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Done, thanks for noting.

File: libcxx/include/__bit/countl.h
Comment (philnik777) [inline libcxx/include/__bit/countl.h:64]: Why is this repeating the signature instead of just changing the body? Same for the other changes.

**Review by philnik777 — CHANGES_REQUESTED** [review]

File: libcxx/include/__bit/countl.h
Comment (marcauberer) [reply libcxx/include/__bit/countl.h]: Done.

**Review by philnik777 — APPROVED** [review]

File: libcxx/include/__bit/countl.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countl.h:70]: This builtin may have an optional second argument which is what to evaluate to if the first argument is zero.
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclzg

```suggestion
#if __has_builtin(__builtin_clzg)
  return __builtin_clzg(__t, numeric_limits<_Tp>::digits);
#else  // __has_builtin(__builtin_clzg)
  if (__t == 0)
    return numeric_limits<_Tp>::digits;

```

This way even more code can get deleted in the future.

File: libcxx/include/__bit/countr.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countr.h:48]: https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fctzg

```suggestion
#if __has_builtin(__builtin_ctzg, numeric_limits<_Tp>::digits)
  return __builtin_ctzg(__t);
#else  // __has_builtin(__builtin_ctzg)
  if (__t == 0)
    return numeric_limits<_Tp>::digits;
```

File: libcxx/include/__bit/countr.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countr.h:44]: Sorry, I may have gotten that wrong in my suggestion, but the optional second argument goes to the invocation of the builtin, not `__has_builtin`.

You compiled this locally, right? :wink:

File: libcxx/include/__bit/countr.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countr.h:44]:
```suggestion
#if __has_builtin(__builtin_ctzg)
  return __builtin_ctzg(__t, numeric_limits<_Tp>::digits);
```

File: libcxx/include/__bit/countr.h
Comment (nickdesaulniers) [reply libcxx/include/__bit/countr.h]: my bad on this suggestion!

File: libcxx/include/__bit/countr.h
Comment (marcauberer) [reply libcxx/include/__bit/countr.h:44]: Was just about to do that, along with a rebase ;)

File: libcxx/include/__bit/countr.h
Comment (marcauberer) [reply libcxx/include/__bit/countr.h]: Done.

File: libcxx/include/__bit/countl.h
Comment (nickdesaulniers) [inline libcxx/include/__bit/countl.h:44]: These headers seem to not put whitespace after the `#`. Mind staying consistent in this file (and libcxx/include/__bit/popcount.h)? Then we should be good to go.  Perhaps your editor is trying to auto indent these.

Comment (philnik777) [reply libcxx/include/__bit/countl.h:44]: No, the indentation is correct.

Comment (marcauberer) [reply libcxx/include/__bit/countl.h:44]: No. This is a clang-format thing. The indentation happens because we have nested #if's here.

Comment (nickdesaulniers) [reply libcxx/include/__bit/countl.h:44]: ah, ok LGTM then. Thanks for your work on this patch!

**Review by nickdesaulniers — APPROVED** [review]

**Comment by marcauberer** [comment]
>There is one issue; our CI is not testing with a Clang version that has https://github.com/llvm/llvm-project/pull/86577. I will look at that and let you know when the CI is ready to really test this patch. Hence the "Request changes" status.

@mordante What's the status on this?

**Review by mordante — APPROVED** [review]
I've updated the CI and it now runs `Ubuntu clang version 19.0.0 (++20240409031520+c8917048e3aa-1~exp1~20240409151624.1609)`. Can you rebase the patch to trigger the CI?

LGTM when the CI passes.

**Pull request merged**
